Lib/test/test_list.py

```diff
@@ -196,6 +196,59 @@ def test_preallocation(self):
         self.assertEqual(iter_size, sys.getsizeof(list([0] * 10)))
         self.assertEqual(iter_size, sys.getsizeof(list(range(10))))
 
+    @cpython_only
+    def test_overallocation(self):
+        # bpo-33234: Don't overallocate when initialized from known lengths
+        # bpo-38373: Allows list over-allocation to be zero for some lengths
+        # bpo-43574: Don't overallocate for list-literals
+        sizeof = sys.getsizeof
+
+        # First handle empty list and empty list-literal cases.  Should have no
+        # overallocation, including init from iterable of unknown length.
+        self.assertEqual(sizeof([]), sizeof(list()))
+        self.assertEqual(sizeof([]), sizeof(list(tuple())))
+        self.assertEqual(sizeof([]), sizeof(list(x for x in [])))
+
+        # Must use actual list-literals to test the overallocation behavior of
+        # compiled list-literals as well as those initialized from them.
+        test_literals = [
+                [1],
+                [1,2],
+                [1,2,3],  # Literals of length > 2 are special-cased in compile
+                [1,2,3,4],
+                [1,2,3,4,5,6,7],
+                [1,2,3,4,5,6,7,8], # bpo-38373: Length 8 init won't over-alloc
+                [1,2,3,4,5,6,7,8,9],
+                ]
+
+        overalloc_amts = []
+        for literal in test_literals:
+            # Ensure that both list literals, and lists made from an iterable
+            # of known size use the same amount of allocation. It will be
+            # verified later that no over-allocation occurs for list literals.
```

Comment:
This comment says "verified later", but I see no specific test of no-over-allocation of list literals. There is `SizeofTest` in `test_sys.py` but it only tests two literal lengths.
I would imagine this kind of test:
```python
prev_literal = None
for literal in [[] + test_literals]:
    if prev_literal is not None
        # Allocation of literal N uses one more pointer than allocation of literal N-1
        self.assertEqual(sizeof(literal), sizeof(prev_literal) + struct.calcsize('P'))
    prev_literal = literal
```

Reply:
Yes, I should have said "verified below", ie. in same test, after the loop.  It's an implicit test, though, since with the current over-allocation strategy, the list capacity can sometimes equal list length, so it's not universally true that each list-literal will have less memory allocated than a normal list of equivalent length.  So it only can test that at least sometimes the list-literals are created with less capacity then a grown list of equivalent length.

I was avoiding a direct test of memory usage, like you propose, just to keep things general.  But I can certainly add such a direct test, which (afaict) is really the only way to test that all the list-literals are being created without over-allocation.

Reply:
@pitrou I ran with your idea, and added a direct test of the memory allocated for a list-literal, which also let me remove the confusing wording about verifying "later", since that's no longer the case.  Thanks for the suggestion.